src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/RuleSetEventHandler.cs

```diff
@@ -0,0 +1,213 @@
+        }
+
+        public void Register()
+        {
+            if (!_eventsHookedUp)
+            {
+                var trackProjectDocuments = (IVsTrackProjectDocuments2)_serviceProvider.GetService(typeof(SVsTrackProjectDocuments));
+
+                if (ErrorHandler.Succeeded(trackProjectDocuments.AdviseTrackProjectDocumentsEvents(this, out _cookie)))
+                {
+                    _eventsHookedUp = true;
+                }
+            }
+        }
+
+        public void Unregister()
+        {
+            if (_eventsHookedUp)
+            {
+                var trackProjectDocuments = (IVsTrackProjectDocuments2)_serviceProvider.GetService(typeof(SVsTrackProjectDocuments));
+
+                if (ErrorHandler.Succeeded(trackProjectDocuments.UnadviseTrackProjectDocumentsEvents(_cookie)))
+                {
+                    _eventsHookedUp = false;
+                    _cookie = 0;
+                }
+            }
+        }
+
+        int IVsTrackProjectDocumentsEvents2.OnAfterAddDirectoriesEx(int cProjects, int cDirectories, IVsProject[] rgpProjects, int[] rgFirstIndices, string[] rgpszMkDocuments, VSADDDIRECTORYFLAGS[] rgFlags)
+        {
+            return VSConstants.S_OK;
+        }
+
+        int IVsTrackProjectDocumentsEvents2.OnAfterAddFilesEx(int cProjects, int cFiles, IVsProject[] rgpProjects, int[] rgFirstIndices, string[] rgpszMkDocuments, VSADDFILEFLAGS[] rgFlags)
+        {
+            return VSConstants.S_OK;
+        }
+
+        int IVsTrackProjectDocumentsEvents2.OnAfterRemoveDirectories(int cProjects, int cDirectories, IVsProject[] rgpProjects, int[] rgFirstIndices, string[] rgpszMkDocuments, VSREMOVEDIRECTORYFLAGS[] rgFlags)
+        {
+            return VSConstants.S_OK;
+        }
+
+        int IVsTrackProjectDocumentsEvents2.OnAfterRemoveFiles(int cProjects, int cFiles, IVsProject[] rgpProjects, int[] rgFirstIndices, string[] rgpszMkDocuments, VSREMOVEFILEFLAGS[] rgFlags)
+        {
+            for (int i = 0; i < rgpProjects.Length; i++)
+            {
+                int indexOfFirstDocumentInProject = IndexOfFirstDocumentInProject(i, rgFirstIndices);
+                int indexOfFirstDocumentInNextProject = IndexOfFirstDocumentInProject(i + 1, rgFirstIndices);
+                for (int j = indexOfFirstDocumentInProject; j < indexOfFirstDocumentInNextProject; j++)
+                {
+                    string fileFullPath = rgpszMkDocuments[j];
+                    if (Path.GetExtension(fileFullPath).Equals(".ruleset", StringComparison.OrdinalIgnoreCase))
+                    {
+                        EnvDTE.Project project;
+                        IVsHierarchy hierarchy = rgpProjects[i] as IVsHierarchy;
+                        if (hierarchy != null &&
+                            hierarchy.TryGetProject(out project))
+                        {
+                            UpdateCodeAnalysisRuleSetPropertiesInProject(project, fileFullPath, string.Empty);
```

Comment:
Yes, we fix up the other projects. If two project use the same file on disk, the file will appear in the list of deleted/renamed files twice, once per project.